common/ManagementTestShared/ManagementRecordTestBase.cs

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Azure.Core;
+using Azure.Core.TestFramework;
+using Azure.ResourceManager.Resources;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using NUnit.Framework;
+using Azure.ResourceManager.Resources.Models;
+using System.Collections.Generic;
+
+
+namespace Azure.ResourceManager.TestFramework
+{
+    [Category("Recorded")]
+    public abstract class ManagementRecordTestBase<TEnvironment> : ClientTestBase where TEnvironment : TestEnvironment, new()
+    {
+        protected RecordedTestSanitizer Sanitizer { get; set; }
+
+        protected RecordMatcher Matcher { get; set; }
+
+        public TEnvironment TestEnvironment { get; }
+
+        private ResourceGroupCleanupPolicy CleanupPolicy { get; set; }
+
+        private uint numberOfResourceGroups;
+
+        protected List<string> resourceGroups;
+
+        protected string defaultResourceGroup;
+
+        private string defaultRgPrefix;
+        protected ManagementRecordTestBase(bool isAsync, string defaultRgPrefix = "Default-", uint numberOfResourceGroups = 1) : this(isAsync, RecordedTestUtilities.GetModeFromEnvironment())
+        {
+            this.numberOfResourceGroups = numberOfResourceGroups;
+            this.resourceGroups = new List<string>();
+            defaultResourceGroup = null;
+            this.defaultRgPrefix = defaultRgPrefix;
+        }
+        protected ManagementRecordTestBase(bool isAsync, RecordedTestMode mode) : base(isAsync)
+        {
+            Sanitizer = new RecordedTestSanitizer();
+            Matcher = new RecordMatcher();
+            Mode = mode;
+            TestEnvironment = new TEnvironment();
+            TestEnvironment.Mode = Mode;
+            CleanupPolicy = new ResourceGroupCleanupPolicy();
+        }
+
+        /// <summary>
+        /// Initializes the clients for a test scenario so the recordings are setup and saved
+        /// correctly
+        /// </summary>
+        protected abstract void InitializeClients();
+
+        /// <summary>
+        /// Setup of resources that will be run before a test fixture or suite
+        /// This should be overriden if resources need to be setup once for a suite of test
+        /// </summary>
+        protected virtual Task AfterOneTimeSetupAsync()
+        {
+            this.InitializeClients();
+            if (this.numberOfResourceGroups < 1)
+                return Task.FromResult<object>(null);
+            this.defaultResourceGroup = Recording.GenerateAssetName(this.defaultRgPrefix);
+            var tmpClient = this.GetResourceManagementClient();
+            tmpClient.ResourceGroups.CreateOrUpdateAsync(defaultResourceGroup, new ResourceGroup("East US"));
```

Comment:
I know we plan to add support for our resource provisioning scripts later but I thought there was a plan to still add the "DeleteAfter" tags so it could still follow our clean-up strategy if for some reason this cannot be cleaned up in the teardown.